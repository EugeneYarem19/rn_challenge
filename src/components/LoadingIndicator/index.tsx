import React from "react";
import { ActivityIndicator, ActivityIndicatorProps, } from "react-native";

import { palette, } from "@theme";

export const LoadingIndicator: React.FC<ActivityIndicatorProps> = ({ color = palette.accentColor, size = "large", }) => (
  <ActivityIndicator size={size} color={color} />
);
