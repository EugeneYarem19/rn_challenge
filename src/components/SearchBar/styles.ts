import { StyleSheet, } from "react-native";

import { palette, } from "@theme";

export const styles = StyleSheet.create({
  container: {
    backgroundColor: palette.transparent,
    borderBottomWidth: 0,
    borderTopWidth: 0,
    borderWidth: 0,
    height: 45,
    marginRight: 15,
    padding: 0,
    width: 235,
  },
});
