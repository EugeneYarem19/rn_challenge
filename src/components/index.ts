export * from "./LoadingIndicator";
export * from "./SearchBar";
