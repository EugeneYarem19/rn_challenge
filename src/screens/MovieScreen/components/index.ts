export * from "./DetailsItem";
export * from "./MovieScreenComponent";
export * from "./Rating";
