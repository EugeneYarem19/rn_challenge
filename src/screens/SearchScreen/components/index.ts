export * from "./SearchResultItem";
export * from "./SearchScreen";
