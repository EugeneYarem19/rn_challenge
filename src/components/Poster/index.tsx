import React from "react";
import ProgressImage from "react-native-image-progress";
import { Icon, } from "react-native-material-ui";
import { View, } from "react-native";

import { LoadingIndicator, } from "../LoadingIndicator";

import { IPoster, ImageSize, } from "./types";
import { styles, } from "./styles";
import { usePoster, } from "./hooks";

export const Poster: React.FC<IPoster> = ({
  poster,
  size = ImageSize.small,
  smallOnEmptyPoster = true,
  smallOnError = true,
  style,
}) => {
  const sizeStyle = size === ImageSize.small ? styles.small : styles.large;
  const { loadError, onError, } = usePoster();

  const renderPlaceholder = (additionalSizedStyle: typeof style) => (
    <View style={[styles.placeholder, additionalSizedStyle, style,]}>
      <Icon name="wallpaper" size={50} />
    </View>
  );

  const renderPosterIcon = () => {
    if (loadError) {
      return renderPlaceholder(smallOnError ? styles.small : sizeStyle);
    } else if (!poster) {
      return renderPlaceholder(smallOnEmptyPoster ? styles.small : sizeStyle);
    } else {
      return (
        <ProgressImage
          source={{ uri: poster, }}
          indicator={LoadingIndicator}
          onError={onError}
          resizeMode="contain"
          style={[styles.poster, sizeStyle, style,]}
          threshold={0}
        />
      );
    }
  };

  return renderPosterIcon();
};
