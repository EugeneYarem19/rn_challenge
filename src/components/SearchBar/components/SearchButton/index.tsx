import React from "react";
import { View, } from "react-native";
import { Button, } from "react-native-material-ui";

import { styles, } from "./styles";

interface SearchCallback {
  (): void;
}

interface Props {
  onPress: SearchCallback;
}

export const SearchButton: React.FC<Props> = ({ onPress, }): JSX.Element => (
  <View testID="SearchButton">
    <Button icon="search" text="" onPress={onPress} style={{ container: styles.container, }} />
  </View>
);
