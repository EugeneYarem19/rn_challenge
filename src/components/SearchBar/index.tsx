import React, { useCallback, useState, } from "react";
import { Keyboard, } from "react-native";
import { SearchBar as SearchBarComponent, } from "react-native-elements";
import { useDispatch, } from "react-redux";

import { actions, } from "@redux";
import { displayInfo, } from "@utils";

import { SearchButton, } from "./components";
import { styles, } from "./styles";

export const SearchBar: React.FC = (): JSX.Element => {
  const dispatch = useDispatch();

  const [searchTitle, setSearchTitle,] = useState("");

  const search = useCallback(() => {
    if (!searchTitle || !searchTitle.length || !searchTitle.trim()) {
      displayInfo("Please, provide movie title to search");
      return;
    }

    Keyboard.dismiss();
    dispatch(actions.findMovies(searchTitle));
  }, [dispatch, searchTitle,]);
  const updateSearchTitle = useCallback((title) => setSearchTitle(title), [setSearchTitle,]);
  const clearSearchTitle = useCallback(() => setSearchTitle(""), [setSearchTitle,]);

  return (
    <SearchBarComponent
      placeholder="Search"
      value={searchTitle}
      onChangeText={updateSearchTitle}
      onClear={clearSearchTitle}
      searchIcon={() => <SearchButton onPress={search} />}
      containerStyle={styles.container}
    />
  );
};
