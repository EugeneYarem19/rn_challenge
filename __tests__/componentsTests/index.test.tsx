import React from "react";
import configureStore from "redux-mock-store";
import { Provider, } from "react-redux";
import { render, fireEvent, } from "@testing-library/react-native";

import { LoadingIndicator, Poster, SearchBar, } from "@components";
import { SearchButton, } from "@src/components/SearchBar/components";

import snapshotTest from "../snapshotTestUtil";
import { mockedActions, mockedData, } from "../mockedData";

const mockStore = configureStore([]);

describe("components tests", () => {
  test("LoadingIndicator must renders correctly", () => {
    snapshotTest(<LoadingIndicator />);
  });

  // tests on the default props is not changed
  test("Poster must renders correctly with set poster", () => {
    snapshotTest(<Poster poster={mockedData.movies[0].poster} />);
  });

  test("Poster must renders correctly without set poster", () => {
    snapshotTest(<Poster poster="" />);
  });

  // tests on different props
  test("Poster must renders correctly with set poster and size = 'small'", () => {
    snapshotTest(<Poster poster={mockedData.movies[0].poster} size="small" />);
  });

  test("Poster must renders correctly with set poster and size = 'large'", () => {
    snapshotTest(<Poster poster={mockedData.movies[0].poster} size="large" />);
  });

  test("Poster must renders correctly without set poster and size = 'small'", () => {
    snapshotTest(<Poster poster="" size="small" />);
  });

  test("Poster must renders correctly without set poster and size = 'small' and smallOnEmptyPoster = 'false'", () => {
    snapshotTest(<Poster poster="" size="small" smallOnEmptyPoster={false} />);
  });

  test("Poster must renders correctly without set poster and size = 'large'", () => {
    snapshotTest(<Poster poster="" size="large" />);
  });

  test("Poster must renders correctly without set poster and size = 'large' and smallOnEmptyPoster = 'false'", () => {
    snapshotTest(<Poster poster="" size="large" smallOnEmptyPoster={false} />);
  });

  test("SearchButton must renders correctly", () => {
    snapshotTest(<SearchButton onPress={() => null} />);
  });

  test("SearchBar must renders correctly", () => {
    const store = mockStore({});
    snapshotTest(
      <Provider store={store}>
        <SearchBar />
      </Provider>
    );
  });

  test("SearchButton must be clickable", () => {
    const onPress = jest.fn();
    const { getByText, } = render(<SearchButton onPress={onPress} />);

    fireEvent.press(getByText(""));

    expect(onPress).toHaveBeenCalled();
  });

  test("SearchBar must start search", () => {
    const store = mockStore({});
    const { getByPlaceholderText, getByTestId, } = render(
      <Provider store={store}>
        <SearchBar />
      </Provider>
    );

    fireEvent.changeText(getByPlaceholderText("Search"), mockedData.title);
    fireEvent.press(getByTestId("SearchButton"));

    const actions = store.getActions();
    const expectedPayload = mockedActions.findMoviesAction;
    expect(actions).toEqual([expectedPayload,]);
  });
});
