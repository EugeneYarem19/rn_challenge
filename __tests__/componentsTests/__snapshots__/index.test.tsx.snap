// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components tests LoadingIndicator must renders correctly 1`] = `
<ActivityIndicator
  animating={true}
  color="#e59f2b"
  hidesWhenStopped={true}
  size="large"
/>
`;

exports[`components tests SearchBar must renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "transparent",
      "borderBottomColor": "#000",
      "borderBottomWidth": 0,
      "borderTopColor": "#000",
      "borderTopWidth": 0,
      "borderWidth": 0,
      "height": 45,
      "marginRight": 15,
      "padding": 0,
      "width": 235,
    }
  }
>
  <View
    style={
      Object {
        "paddingHorizontal": 0,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#303337",
          "borderBottomWidth": 0,
          "borderColor": "#86939e",
          "borderRadius": 3,
          "flexDirection": "row",
          "minHeight": 30,
          "overflow": "hidden",
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 40,
            "justifyContent": "center",
            "marginLeft": 8,
            "marginVertical": 4,
            "paddingRight": 4,
          }
        }
      >
        <View
          testID="SearchButton"
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onLayout={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            pointerEvents="box-only"
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderRadius": 2,
                  "flexDirection": "row",
                  "height": 36,
                  "justifyContent": "center",
                  "paddingHorizontal": 16,
                },
                undefined,
                false,
                false,
                false,
                Object {},
                Object {
                  "backgroundColor": "#e59f2b",
                  "borderWidth": 1,
                  "height": 35,
                  "padding": 0,
                  "width": 60,
                },
              ]
            }
            testID={null}
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontSize": 24,
                  },
                  Array [
                    Object {
                      "marginRight": 8,
                    },
                    undefined,
                    false,
                    false,
                    undefined,
                    undefined,
                  ],
                  Object {
                    "fontFamily": "Material Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontSize": 14,
                    "fontWeight": "500",
                  },
                  undefined,
                  false,
                  false,
                  false,
                  undefined,
                  undefined,
                ]
              }
            >
              
            </Text>
            <View
              pointerEvents="none"
              style={
                Array [
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "overflow": "hidden",
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  undefined,
                ]
              }
            >
              <View
                pointerEvents="none"
                style={
                  Object {
                    "backgroundColor": "rgba(0, 0, 0, 0.87)",
                    "bottom": 0,
                    "left": 0,
                    "opacity": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "zIndex": 1,
                  }
                }
              />
              <View
                pointerEvents="none"
                style={
                  Object {
                    "backgroundColor": "rgba(0, 0, 0, 0.87)",
                    "borderRadius": 100,
                    "height": 200,
                    "left": -100,
                    "opacity": 0.12,
                    "position": "absolute",
                    "top": -100,
                    "transform": Array [
                      Object {
                        "scale": 0,
                      },
                    ],
                    "width": 200,
                    "zIndex": 1,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <TextInput
        allowFontScaling={true}
        editable={true}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Search"
        placeholderTextColor="#86939e"
        platform="default"
        rejectResponderTermination={true}
        style={
          Object {
            "alignSelf": "center",
            "color": "#86939e",
            "flex": 1,
            "fontSize": 18,
            "marginLeft": 10,
            "minHeight": 40,
          }
        }
        testID="SearchBar"
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 40,
            "justifyContent": "center",
            "marginRight": 8,
            "marginVertical": 4,
            "paddingRight": 4,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        />
      </View>
    </View>
    <Text
      style={
        Object {
          "color": "#ff190c",
          "fontSize": 12,
          "height": 0,
          "margin": 0,
          "padding": 0,
        }
      }
    />
  </View>
</View>
`;

exports[`components tests SearchButton must renders correctly 1`] = `
<View
  testID="SearchButton"
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onLayout={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    pointerEvents="box-only"
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 2,
          "flexDirection": "row",
          "height": 36,
          "justifyContent": "center",
          "paddingHorizontal": 16,
        },
        undefined,
        false,
        false,
        false,
        Object {},
        Object {
          "backgroundColor": "#e59f2b",
          "borderWidth": 1,
          "height": 35,
          "padding": 0,
          "width": 60,
        },
      ]
    }
    testID={null}
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#000000",
            "fontSize": 24,
          },
          Array [
            Object {
              "marginRight": 8,
            },
            undefined,
            false,
            false,
            undefined,
            undefined,
          ],
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontSize": 14,
            "fontWeight": "500",
          },
          undefined,
          false,
          false,
          false,
          undefined,
          undefined,
        ]
      }
    >
      
    </Text>
    <View
      pointerEvents="none"
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "overflow": "hidden",
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          undefined,
        ]
      }
    >
      <View
        pointerEvents="none"
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.87)",
            "bottom": 0,
            "left": 0,
            "opacity": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 1,
          }
        }
      />
      <View
        pointerEvents="none"
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.87)",
            "borderRadius": 100,
            "height": 200,
            "left": -100,
            "opacity": 0.12,
            "position": "absolute",
            "top": -100,
            "transform": Array [
              Object {
                "scale": 0,
              },
            ],
            "width": 200,
            "zIndex": 1,
          }
        }
      />
    </View>
  </View>
</View>
`;
